AssertionError: duplicate type when exporting conflicting type members

## minimized code

```Scala
class A{
  type T = Int // can also be class T
}
class B(x: A, y: A){
  export x._
  export y._
}
```

## Crash output (click arrow to expand)
<details>

```
exception occurred while typechecking export-type-conflict.scala
exception occurred while compiling export-type-conflict.scala
java.lang.AssertionError: assertion failed: duplicate type T#3901; previous was type T#3900 while compiling export-type-conflict.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: duplicate type T#3901; previous was type T#3900
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
	at dotty.tools.dotc.core.Scopes$MutableScope.enter(Scopes.scala:269)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.enterNoReplace(SymDenotations.scala:1763)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.enter(SymDenotations.scala:1752)
	at dotty.tools.dotc.core.Symbols$Symbol.entered(Symbols.scala:571)
	at dotty.tools.dotc.typer.Namer.processExports$$anonfun$2$$anonfun$1(Namer.scala:1151)
	at scala.collection.immutable.List.foreach(List.scala:305)
	at dotty.tools.dotc.typer.Namer.dotty$tools$dotc$typer$Namer$ClassCompleter$$_$processExports$$anonfun$1(Namer.scala:1151)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:305)
	at dotty.tools.dotc.typer.Namer$ClassCompleter.processExports(Namer.scala:1151)
	at dotty.tools.dotc.typer.Namer$ClassCompleter.completeInCreationContext(Namer.scala:1291)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:840)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:260)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:185)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:187)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:400)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2118)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2143)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2223)
	at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2262)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2271)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2274)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2296)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2340)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1959)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2197)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2224)
	at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2262)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2271)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2274)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2382)
	at dotty.tools.dotc.typer.FrontEnd.liftedTree1$2(FrontEnd.scala:78)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:83)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:42)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:84)
	at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:114)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:305)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:114)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:167)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:177)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:185)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:192)
	at dotty.tools.dotc.Run.compileSources(Run.scala:129)
	at dotty.tools.dotc.Run.compile(Run.scala:112)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:36)
	at dotty.tools.dotc.Driver.process(Driver.scala:189)
	at dotty.tools.dotc.Driver.process(Driver.scala:158)
	at dotty.tools.dotc.Driver.process(Driver.scala:170)
	at dotty.tools.dotc.Driver.main(Driver.scala:197)
	at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -265,6 +265,49 @@ class Namer { typer: Typer =>
     sym
   }
 
+  /** Check that a new definition with given name and privacy status
+   *  in current context would not conflict with existing currently
+   *  compiled definitions.
+   *  The logic here is very subtle and fragile due to the fact that
+   *  we are not allowed to force anything.
+   */
+  def checkNoConflict(name: Name, isPrivate: Boolean, span: Span)(using ctx: Context): Name =
+    val owner = ctx.owner
+    var conflictsDetected = false
+
+    def conflict(conflicting: Symbol) =
+      val where: String =
+        if conflicting.owner == owner then ""
+        else if conflicting.owner.isPackageObject then i" in ${conflicting.associatedFile}"
+        else i" in ${conflicting.owner}"
+      ctx.error(i"$name is already defined as $conflicting$where", ctx.source.atSpan(span))
+      conflictsDetected = true
+
+    def checkNoConflictIn(owner: Symbol) =
+      val preExisting = owner.unforcedDecls.lookup(name)
+      if (preExisting.isDefinedInCurrentRun || preExisting.lastKnownDenotation.is(Package))
+          && (!preExisting.lastKnownDenotation.is(Private) || preExisting.owner.is(Package))
+      then conflict(preExisting)
+
+    def pkgObjs(pkg: Symbol) =
+      pkg.denot.asInstanceOf[PackageClassDenotation].packageObjs.map(_.symbol)
+
+    if owner.is(PackageClass) then
+      checkNoConflictIn(owner)
+      for pkgObj <- pkgObjs(owner) do
+        checkNoConflictIn(pkgObj)
+    else
+      def preExisting = ctx.effectiveScope.lookup(name)
+      if (!owner.isClass || name.isTypeName) && preExisting.exists then
+        conflict(preExisting)
+      else if owner.isPackageObject && !isPrivate && name != nme.CONSTRUCTOR then
+        checkNoConflictIn(owner.owner)
+        for pkgObj <- pkgObjs(owner.owner) if pkgObj != owner do
+          checkNoConflictIn(pkgObj)
+
+    if conflictsDetected then name.freshened else name
+  end checkNoConflict
+
   /** If this tree is a member def or an import, create a symbol of it
    *  and store in symOfTree map.
    */
@@ -302,49 +345,6 @@ class Namer { typer: Typer =>
 
     typr.println(i"creating symbol for $tree in ${ctx.mode}")
 
-    /** Check that a new definition with given name and privacy status
-     *  in current context would not conflict with existing currently
-     *  compiled definitions.
-     *  The logic here is very subtle and fragile due to the fact that
-     *  we are not allowed to force anything.
-     */
-    def checkNoConflict(name: Name, isPrivate: Boolean): Name =
-      val owner = ctx.owner
-      var conflictsDetected = false
-
-      def conflict(conflicting: Symbol) =
-        val where: String =
-          if conflicting.owner == owner then ""
-          else if conflicting.owner.isPackageObject then i" in ${conflicting.associatedFile}"
-          else i" in ${conflicting.owner}"
-        ctx.error(i"$name is already defined as $conflicting$where", tree.sourcePos)
-        conflictsDetected = true
-
-      def checkNoConflictIn(owner: Symbol) =
-        val preExisting = owner.unforcedDecls.lookup(name)
-        if (preExisting.isDefinedInCurrentRun || preExisting.lastKnownDenotation.is(Package))
-           && (!preExisting.lastKnownDenotation.is(Private) || preExisting.owner.is(Package))
-        then conflict(preExisting)
-
-      def pkgObjs(pkg: Symbol) =
-        pkg.denot.asInstanceOf[PackageClassDenotation].packageObjs.map(_.symbol)
-
-      if owner.is(PackageClass) then
-        checkNoConflictIn(owner)
-        for pkgObj <- pkgObjs(owner) do
-          checkNoConflictIn(pkgObj)
-      else
-        def preExisting = ctx.effectiveScope.lookup(name)
-        if (!owner.isClass || name.isTypeName) && preExisting.exists then
-          conflict(preExisting)
-        else if owner.isPackageObject && !isPrivate && name != nme.CONSTRUCTOR then
-          checkNoConflictIn(owner.owner)
-          for pkgObj <- pkgObjs(owner.owner) if pkgObj != owner do
-            checkNoConflictIn(pkgObj)
-
-      if conflictsDetected then name.freshened else name
-    end checkNoConflict
-
     /** Create new symbol or redefine existing symbol under lateCompile. */
     def createOrRefine[S <: Symbol](
         tree: MemberDef, name: Name, flags: FlagSet, owner: Symbol, infoFn: S => Type,
@@ -376,7 +376,7 @@ class Namer { typer: Typer =>
     tree match {
       case tree: TypeDef if tree.isClassDef =>
         val flags = checkFlags(tree.mods.flags &~ GivenOrImplicit)
-        val name = checkNoConflict(tree.name, flags.is(Private)).asTypeName
+        val name = checkNoConflict(tree.name, flags.is(Private), tree.span).asTypeName
         val cls =
           createOrRefine[ClassSymbol](tree, name, flags, ctx.owner,
             cls => adjustIfModule(new ClassCompleter(cls, tree)(ctx), tree),
@@ -385,7 +385,7 @@ class Namer { typer: Typer =>
         cls
       case tree: MemberDef =>
         var flags = checkFlags(tree.mods.flags)
-        val name = checkNoConflict(tree.name, flags.is(Private))
+        val name = checkNoConflict(tree.name, flags.is(Private), tree.span)
         tree match
           case tree: ValOrDefDef =>
             if tree.unforcedRhs == EmptyTree
@@ -1079,9 +1079,10 @@ class Namer { typer: Typer =>
           if (whyNoForwarder(mbr) == "") {
             val sym = mbr.symbol
             val forwarder =
-              if (mbr.isType)
+              if mbr.isType then
+                val forwarderName = checkNoConflict(alias.toTypeName, isPrivate = false, span)
                 ctx.newSymbol(
-                  cls, alias.toTypeName,
+                  cls, forwarderName,
                   Exported | Final,
                   TypeAlias(path.tpe.select(sym)),
                   coord = span)
@@ -1096,7 +1097,8 @@ class Namer { typer: Typer =>
                   else
                     (EmptyFlags, mbr.info.ensureMethodic)
                 val mbrFlags = Exported | Method | Final | maybeStable | sym.flags & RetainedExportFlags
-                ctx.newSymbol(cls, alias, mbrFlags, mbrInfo, coord = span)
+                val forwarderName = checkNoConflict(alias, isPrivate = false, span)
+                ctx.newSymbol(cls, forwarderName, mbrFlags, mbrInfo, coord = span)
               }
             forwarder.info = avoidPrivateLeaks(forwarder)
             val forwarderDef =
@@ -1146,9 +1148,9 @@ class Namer { typer: Typer =>
         forwarders
       }
 
-      val forwarderss =
-        for (exp @ Export(_, _) <- rest) yield exportForwarders(exp)
-      forwarderss.foreach(_.foreach(fwdr => fwdr.symbol.entered))
+      for case exp @ Export(_, _) <- rest do
+        for forwarder <- exportForwarders(exp) do
+          forwarder.symbol.entered
     }
 
     /** Ensure constructor is completed so that any parameter accessors

--- a/tests/neg/i8333.scala
+++ b/tests/neg/i8333.scala
@@ -1,0 +1,13 @@
+class A:
+  type T = Int // can also be class T
+class B(x: A, y: A):
+  export x._
+  export y._  // error: duplicate
+class C(x: A):
+  type T = String
+  export x._  // error: duplicate
+class D(x: A):
+  export x._  // error: duplicate
+  type T = String
+
+

